common/sealable-trie/src/bits.rs

```diff
@@ -440,39 +434,41 @@ impl<'a> Slice<'a> {
         (ExtKey::try_from(prefix).ok(), ExtKey::try_from(suffix).ok())
     }
 
-    /// Returns total number of underlying bits, i.e. bits in the slice plus the
-    /// offset.
-    fn underlying_bits_length(&self) -> usize {
-        usize::from(self.offset) + usize::from(self.length)
-    }
-
     /// Returns bytes underlying the bit slice.
+    #[inline]
     fn bytes(&self) -> &'a [u8] {
-        // We need to special-case zero length to make sure that in situation of
-        // non-zero offset and zero length we return an empty slice.
-        let len = match self.length {
-            0 => 0,
-            _ => (self.underlying_bits_length() + 7) / 8,
-        };
         // SAFETY: `ptr` is guaranteed to be valid pointer point at `offset +
         // length` valid bits.
-        unsafe { core::slice::from_raw_parts(self.ptr, len) }
+        unsafe { core::slice::from_raw_parts(self.ptr, self.bytes_len()) }
     }
 
+    /// Calculates underlying bytes length of the slice.
+    #[inline]
+    fn bytes_len(&self) -> usize { bytes_len(self.offset, self.length) }
+
     /// Helper method which returns masks for leading and trailing byte.
     ///
     /// Based on provided bit offset (which must be ≤ 7) and bit length of the
     /// slice returns: mask of bits in the first byte that are part of the
     /// slice and mask of bits in the last byte that are part of the slice.
+    #[inline]
     fn masks(offset: U3, length: u16) -> (u8, u8) {
-        let bits = usize::from(offset) + usize::from(length);
-        // `1 << 20` is an arbitrary number which is divisible by 8 and greater
-        // than bits.
-        let tail = ((1 << 20) - bits) % 8;
+        let tail = -offset.wrapping_add(length);
         (0xFFu8 >> offset, 0xFFu8 << tail)
     }
 }
 
+/// Calculates underlying bytes length of a slice with given offset and length.
+#[inline]
+fn bytes_len(offset: U3, length: u16) -> usize {
+    // We need to special-case zero length to make sure that in situation of
+    // non-zero offset and zero length we return an empty slice.
+    match length {
+        0 => 0,
+        _ => ((u32::from(offset) + u32::from(length) + 7) / 8) as usize,
```

Comment:
is the numerator always a multiple of 8? does it matter at all?

Reply:
No, numerator can be arbitrary number. This divides by eight with rounding up to calculate number of bytes needed for given number of bits.  That is, if `offset + length` is say 10 (bits) that occupies 2 (bytes).